include/surface_perception/surface_finder.h

```diff
@@ -0,0 +1,133 @@
+#include <vector>
+
+#include "pcl/ModelCoefficients.h"
+#include "pcl/PointIndices.h"
+#include "pcl/point_cloud.h"
+#include "pcl/point_types.h"
+
+namespace surface_perception {
+/// \brief SurfaceFinder attempt to find multiple horizontal surfaces given a
+/// shelf scene.
+///
+/// This class is designed to find surfaces in a shelf scene such as a
+/// bookshelf. In particular, this class finds the horizontal surfaces if the
+/// following conditions are met:
+/// 1. No NaN points in the input cloud
+/// 2. Each target surface has different height
+///
+/// If the input cloud meets the requirement. Mutation functions of this class
+/// can be used to adjust the parameters based on the scenario of the point
+/// cloud scene.
+///
+/// \b Example usage:
+/// \code
+///   SurfaceFinder finder;
+///   finder.set_cloud(pcl_cloud);
+///   finder.set_cloud_indices(point_indices);
+///   finder.set_angle_tolerance(10);
+///   finder.set_max_point_distance(0.01);
+///   finder.set_min_iteration(1000);
+///   finder.set_surface_point_threshold(10000);
+///
+///   std::vector<pcl::PointIndices::Ptr> indices;
+///   std::vector<pcl::ModelCoefficients> coeffs;
+///   finder.ExploreSurfaces(0, 10, &indices, &coeffs);
+/// \endcode
+class SurfaceFinder {
+ public:
+  /// \brief Default constructor
+  SurfaceFinder();
+
+  /// \brief Set the input point cloud.
+  ///
+  /// \param[in] cloud 	The input cloud for surface detection. NaN values in
+  ///  the input cloud should be removed before being passed to this function.
+  void set_cloud(const pcl::PointCloud<pcl::PointXYZRGB>::Ptr& cloud);
+
+  /// \brief Set the indices of the input point cloud.
+  ///
+  /// \param[in] cloud_indices The indices of the input point cloud.
+  void set_cloud_indices(const pcl::PointIndices::Ptr cloud_indices);
+
+  /// \brief Set the angle of the surface compare to horizontal surface
+  ///
+  /// Because the current state algorithm only search for horizontal surfaces,
+  /// the angle set through this function is not used in surface detection.
+  ///
+  /// \param[in] angle_tolerance The maximum angle difference between a
+  ///  surface candidate against a horizontal surface.
+  void set_angle_tolerance(const double& angle_tolerance);
+
+  /// \brief Set the maximum distance for a point to be considered part of
+  /// surfaces.
+  ///
+  /// \param[in] max_point_distance The maximum distance between a point and
+  ///  a plane that represents the surface.
+  void set_max_point_distance(const double& max_point_distance);
+
+  /// \brief Set the minimum number of iterations for the algorithm to find
+  /// surfaces
+  ///
+  /// The algorithm is designed to run at least the given number of iteration or
+  /// finding the specified minimum number of surfaces in a given scene.
+  ///
+  /// \param[in] min_iteration The minimum number of iteration required before
+  ///  the algorithm can stop search for surfaces.
+  void set_min_iteration(const size_t& min_iteration);
```

Comment:
I think int is fine. I doubt we want values greater than 2 billion.

Reply:
Completed

Reply:
Revert back to size_t without being constant reference.